src/server/middleware/vulnerable/vulnerable.go

```diff
@@ -1,125 +1,145 @@
-					serror.SendError(rw, pkgE)
-					return
-				}
+		securityCtx, ok := security.FromContext(ctx)
+		if ok &&
+			securityCtx.Name() == "robot" &&
+			securityCtx.Can(rbac.ActionScannerPull, rbac.NewProjectNamespace(proj.ProjectID).Resource(rbac.ResourceRepository)) {
+			// the artifact is pulling by the scanner, skip the checking
+			logger.Debugf("artifact %s@%s is pulling by the scanner, skip the checking", art.RepositoryName, art.Digest)
+			return nil
+		}
 
-				rawSummary, ok := summaries[v1.MimeTypeNativeReport]
-				// No report yet?
-				if !ok {
-					err = errors.Errorf("no scan report existing for the artifact: %s:%s@%s", img.Repository, img.Tag, img.Digest)
-					pkgE := internal_errors.New(err).WithCode(internal_errors.PROJECTPOLICYVIOLATION)
-					serror.SendError(rw, pkgE)
-					return
-				}
+		checker := scanChecker()
 
-				summary := rawSummary.(*vuln.NativeReportSummary)
+		scannable, err := checker.IsScannable(ctx, art)
+		if err != nil {
+			logger.Errorf("check the scannable status of the artifact %s@%s failed, error: %v", art.RepositoryName, art.Digest, err)
+			return err
+		}
 
-				// Do judgement
-				if summary.Severity.Code() >= projectVulnerableSeverity.Code() {
-					err = errors.Errorf("current image with '%q vulnerable' cannot be pulled due to configured policy in 'Prevent images with vulnerability severity of %q from running.' "+
-						"Please contact your project administrator for help'", summary.Severity, projectVulnerableSeverity)
-					pkgE := internal_errors.New(err).WithCode(internal_errors.PROJECTPOLICYVIOLATION)
-					serror.SendError(rw, pkgE)
-					return
-				}
+		if !scannable {
+			// the artifact is not scannable, skip the checking
+			logger.Debugf("artifact %s@%s is not scannable, skip the checking", art.RepositoryName, art.Digest)
+			return nil
+		}
+
+		whitelist := report.CVESet(proj.CVEWhitelist.CVESet())
+		summaries, err := scanController.GetSummary(ctx, art, []string{v1.MimeTypeNativeReport}, report.WithCVEWhitelist(&whitelist))
+		if err != nil {
+			logger.Errorf("get vulnerability summary of the artifact %s@%s failed, error: %v", art.RepositoryName, art.Digest, err)
+			return err
+		}
 
-				// Print scannerPull CVE list
-				if len(summary.CVEBypassed) > 0 {
-					for _, cve := range summary.CVEBypassed {
-						log.Infof("Vulnerable policy check: scannerPull CVE %s", cve)
-					}
+		rawSummary, ok := summaries[v1.MimeTypeNativeReport]
+		if !ok {
+			// No report yet?
+			msg := "vulnerability prevention enabled, but no scan report existing for the artifact"
+			return ierror.New(nil).WithCode(ierror.PROJECTPOLICYVIOLATION).WithMessage(msg)
+		}
+
+		summary, ok := rawSummary.(*vuln.NativeReportSummary)
+		if !ok {
+			return fmt.Errorf("report summary is invalid")
+		}
+
+		hasCapability, err := checker.HasCapability(ctx, art)
+		if err != nil {
+			return err // this will never happened because the artifact is scannable
+		}
+		if !hasCapability {
```

Comment:
I would rather use `art.HasChildren` to figure if it's an index.  Even if a scanner declares it has the capability of scanning an index, we should still treat the index the same way when handling it's scan result in the middleware.